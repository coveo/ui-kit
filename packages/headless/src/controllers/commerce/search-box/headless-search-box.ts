import {CommerceEngine} from '../../../app/commerce-engine/commerce-engine';
import {configuration} from '../../../app/common-reducers';
import {deselectAllBreadcrumbs} from '../../../features/breadcrumb/breadcrumb-actions';
import {selectPage} from '../../../features/commerce/pagination/pagination-actions';
import {fetchQuerySuggestions} from '../../../features/commerce/query-suggest/query-suggest-actions';
import {updateQuery} from '../../../features/commerce/query/query-actions';
import {queryReducer as commerceQuery} from '../../../features/commerce/query/query-slice';
import {executeSearch} from '../../../features/commerce/search/search-actions';
import {commerceSearchReducer as commerceSearch} from '../../../features/commerce/search/search-slice';
import {updateFacetAutoSelection} from '../../../features/facets/generic/facet-actions';
import {
  registerQuerySetQuery,
  updateQuerySetQuery,
} from '../../../features/query-set/query-set-actions';
import {querySetReducer as querySet} from '../../../features/query-set/query-set-slice';
import {
  clearQuerySuggest,
  registerQuerySuggest,
  selectQuerySuggestion,
} from '../../../features/query-suggest/query-suggest-actions';
import {querySuggestReducer as querySuggest} from '../../../features/query-suggest/query-suggest-slice';
import {
  CommerceQuerySection,
  CommerceSearchSection,
  ConfigurationSection,
  QuerySetSection,
  QuerySuggestionSection,
} from '../../../state/state-sections';
import {loadReducerError} from '../../../utils/errors';
import {
  Delimiters,
  SuggestionHighlightingOptions,
} from '../../../utils/highlight';
import {randomID} from '../../../utils/utils';
import {validateOptions} from '../../../utils/validate-payload';
import {buildController} from '../../controller/headless-controller';
import {
  getSuggestions,
  SearchBoxState,
  SearchBox as CoreSearchBox,
  Suggestion,
} from '../../core/search-box/headless-core-search-box';
import {
  defaultSearchBoxOptions,
  SearchBoxOptions,
  searchBoxOptionsSchema,
} from './headless-search-box-options';

export type {
  SearchBoxOptions,
  SearchBoxState,
  SuggestionHighlightingOptions,
  Suggestion,
  Delimiters,
};

export interface SearchBoxProps {
  /**
   * The `SearchBox` controller options.
   */
  options?: SearchBoxOptions;
}

/**
 * The `SearchBox` headless controller offers a high-level interface for designing a common search box UI component
 * with [highlighting for query suggestions](https://docs.coveo.com/en/headless/latest/usage/highlighting/).
 */
export type SearchBox = Omit<CoreSearchBox, 'submit'> & {
  /**
   * Triggers a commerce search query.
   */
  submit(): void;
};

/**
 * Creates a `SearchBox` controller instance.
 *
 * @param engine - The commerce headless engine.
 * @param props - The configurable `SearchBox` properties.
 * @returns A `SearchBox` controller instance.
 */
export function buildSearchBox(
  engine: CommerceEngine,
  props: SearchBoxProps = {}
): SearchBox {
  if (!loadSearchBoxReducers(engine)) {
    throw loadReducerError;
  }

  const controller = buildController(engine);
  const {dispatch} = engine;
  const getState = () => engine.state;

  const id = props.options?.id || randomID('search_box');
  const options: Required<SearchBoxOptions> = {
    id,
    highlightOptions: {...props.options?.highlightOptions},
    ...defaultSearchBoxOptions,
    ...props.options,
  };

  validateOptions(engine, searchBoxOptionsSchema, options, 'buildSearchBox');
  dispatch(
    registerQuerySetQuery({id, query: engine.state.commerceQuery.query ?? ''})
  );
  dispatch(registerQuerySuggest({id}));

  const getValue = () => engine.state.querySet[options.id];

  const performSearch = async () => {
    if (options.clearFilters) {
      dispatch(deselectAllBreadcrumbs());
    }

    dispatch(updateFacetAutoSelection({allow: true}));
    dispatch(updateQuery({query: getValue()}));
    dispatch(selectPage(0));
    dispatch(executeSearch());
  };

  return {
    ...controller,

    updateText(value: string) {
      dispatch(updateQuerySetQuery({id, query: value}));
      this.showSuggestions();
    },

    clear() {
      dispatch(updateQuerySetQuery({id, query: ''}));
      dispatch(clearQuerySuggest({id}));
    },

    showSuggestions() {
      dispatch(fetchQuerySuggestions({id}));
    },

    selectSuggestion(value: string) {
      dispatch(selectQuerySuggestion({id, expression: value}));
      performSearch().then(() => {
        dispatch(clearQuerySuggest({id}));
      });
    },

    submit() {
      performSearch();
      dispatch(clearQuerySuggest({id}));
    },

    get state() {
      const state = getState();
      const querySuggest = state.querySuggest[options.id];
      const suggestions = getSuggestions(
        querySuggest,
        options.highlightOptions
      );
      const isLoadingSuggestions = querySuggest
        ? querySuggest.isLoading
        : false;

      return {
        value: getValue(),
        suggestions,
        isLoading: state.commerceSearch.isLoading,
        isLoadingSuggestions,
      };
    },
  };
}

function loadSearchBoxReducers(
  engine: CommerceEngine
): engine is CommerceEngine<
  CommerceQuerySection &
    QuerySuggestionSection &
    ConfigurationSection &
    QuerySetSection &
    CommerceSearchSection
> {
  engine.addReducers({
    commerceQuery,
    querySuggest,
    configuration,
    querySet,
    commerceSearch,
  });
  return true;
}
