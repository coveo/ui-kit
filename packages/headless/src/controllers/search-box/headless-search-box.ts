import {
  fetchQuerySuggestions,
  clearQuerySuggest,
  clearQuerySuggestCompletions,
  registerQuerySuggest,
  selectQuerySuggestion,
} from '../../features/query-suggest/query-suggest-actions';
import {Engine} from '../../app/headless-engine';
import {updateQuery} from '../../features/query/query-actions';
import {
  registerQuerySetQuery,
  updateQuerySetQuery,
} from '../../features/query-set/query-set-actions';
import {executeSearch} from '../../features/search/search-actions';
import {buildController} from '../controller/headless-controller';
import {updatePage} from '../../features/pagination/pagination-actions';
import {logSearchboxSubmit} from '../../features/query/query-analytics-actions';
import {
  ConfigurationSection,
  QuerySection,
  QuerySetSection,
  QuerySuggestionSection,
  SearchSection,
} from '../../state/state-sections';
import {
  SearchBoxOptions,
  defaultSearchBoxOptions,
  searchBoxOptionsSchema,
} from './headless-search-box-options';
import {validateOptions} from '../../utils/validate-payload';
import {logQuerySuggestionClick} from '../../features/query-suggest/query-suggest-analytics-actions';
import {randomID} from '../../utils/utils';
import {QuerySuggestState} from '../../features/query-suggest/query-suggest-state';
import {SearchAction} from '../../features/analytics/analytics-utils';
import {
  getHighlightedSuggestion,
  SuggestionHighlightingOptions,
} from '../../utils/highlight';

export {SearchBoxOptions};
export interface SearchBoxProps {
  options: SearchBoxOptions;
}

/**
 * A scoped and simplified part of the headless state that is relevant to the `SearchBox` controller.
 */
export type SearchBoxState = SearchBox['state'];
/**
 * The `SearchBox` headless controller offers a high-level interface for designing a common search box UI controller.
 */
export type SearchBox = ReturnType<typeof buildSearchBox>;

export function buildSearchBox(
  engine: Engine<
    QuerySection &
      QuerySuggestionSection &
      ConfigurationSection &
      QuerySetSection &
      SearchSection
  >,
  props: Partial<SearchBoxProps> = {}
) {
  const controller = buildController(engine);
  const {dispatch} = engine;

  const id = props.options?.id || randomID('search_box');
  const options: Required<SearchBoxOptions> = {
    id,
    highlightOptions: {...props.options?.highlightOptions},
    ...defaultSearchBoxOptions,
    ...props.options,
  };

  validateOptions(engine, searchBoxOptionsSchema, options, 'buildSearchBox');

  dispatch(registerQuerySetQuery({id, query: ''}));
  dispatch(
    registerQuerySuggest({
      id,
      q: engine.state.query.q,
      count: options.numberOfSuggestions,
    })
  );

  const getValue = () => engine.state.querySet[options.id];

  const performSearch = (analytics: SearchAction) => {
    const {enableQuerySyntax} = options;

    dispatch(updateQuery({q: getValue(), enableQuerySyntax}));
    dispatch(updatePage(1));
    dispatch(executeSearch(analytics));
  };

  return {
    ...controller,

    /**
     * Updates the search box text value and shows the suggestions for that value.
     * @param value  The string value to update the search box with.
     */
    updateText(value: string) {
      dispatch(updateQuerySetQuery({id, query: value}));
      this.showSuggestions();
    },

    /**
     * Clears the search box text and the suggestions.
     */
    clear() {
      dispatch(updateQuerySetQuery({id, query: ''}));
      dispatch(clearQuerySuggest({id}));
    },

    /**
     * @deprecated suggestions should be hidden using css instead for an optimal user experience.
     * Clears the suggestions.
     */
    hideSuggestions() {
      dispatch(clearQuerySuggestCompletions({id}));
    },

    /**
     * Shows the suggestions for the current search box value.
     */
    showSuggestions() {
      if (options.numberOfSuggestions) {
        dispatch(fetchQuerySuggestions({id}));
      }
    },

    /**
     * Selects a suggestion and calls `submit`.
     * @param value The string value of the suggestion to select
     */
    selectSuggestion(value: string) {
      dispatch(selectQuerySuggestion({id, expression: value}));
      performSearch(logQuerySuggestionClick({id, suggestion: value}));
    },

    /**
     * Triggers a search query.
     */
    submit() {
      performSearch(logSearchboxSubmit());
    },

    /**
     * @returns The state of the `SearchBox` controller.
     */
    get state() {
      const state = engine.state;
      const querySuggestState = state.querySuggest[options.id];
      const suggestions = getSuggestions(
        querySuggestState,
        options.highlightOptions
      );

      return {
        value: getValue(),
        suggestions,
        isLoading: state.search.isLoading,
      };
    },
  };
}

function getSuggestions(
  state: QuerySuggestState | undefined,
  highlightOptions: SuggestionHighlightingOptions
) {
  if (!state) {
    return [];
  }

  return state.completions.map((completion) => ({
    highlightedValue: getHighlightedSuggestion(
      completion.highlighted,
      highlightOptions
    ),
    rawValue: completion.expression,
  }));
}
