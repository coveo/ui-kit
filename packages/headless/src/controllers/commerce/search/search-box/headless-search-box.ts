import {logSearchboxSubmit} from '../../../../features/query/query-analytics-actions';
import {executeSearch, fetchQuerySuggestions} from '../../../../features/commerce/search/search-actions';
import {SuggestionHighlightingOptions, Delimiters} from '../../../../utils/highlight';
import {
  buildCoreSearchBox, SearchBox,
  SearchBoxState,
  Suggestion,
} from '../../../core/search-box/headless-core-search-box';
import {SearchBoxOptions} from '../../../core/search-box/headless-core-search-box-options';
import {CommerceEngine} from '../../../../app/commerce-engine/commerce-engine';

export type {
  SearchBoxOptions,
  SearchBoxState,
  SuggestionHighlightingOptions,
  Suggestion,
  SearchBox,
  Delimiters,
};

export interface SearchBoxProps {
  /**
   * The `SearchBox` controller options.
   */
  options?: SearchBoxOptions;
}

/**
 * Creates a commerce `SearchBox` controller instance.
 *
 * @param engine - The commerce engine.
 * @param props - The configurable `SearchBox` properties.
 * @returns A `SearchBox` controller instance.
 */
export function buildSearchBox(
  engine: CommerceEngine,
  props: SearchBoxProps = {}
): SearchBox {
  const searchBox = buildCoreSearchBox(engine, {
    ...props,
    executeSearchActionCreator: executeSearch,
    fetchQuerySuggestionsActionCreator: fetchQuerySuggestions,
  });

  return {
    ...searchBox,
    submit() {
      searchBox.submit(logSearchboxSubmit());
    },
    get state() {
      return searchBox.state;
    },
  };
}
