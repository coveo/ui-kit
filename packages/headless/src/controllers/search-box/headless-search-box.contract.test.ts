import {PactV3} from '@pact-foundation/pact';
import {describe, it, expect} from 'vitest';
import {
  providerFactory,
  addSuggestionInteraction,
} from '../../../pact-provider/provider.js';
import {
  buildSearchEngine,
  getSampleSearchEngineConfiguration,
} from '../../app/search-engine/search-engine.js';
import {buildSearchBox, SearchBox} from './headless-search-box.js';

const getSearchBoxController = (url: string) => {
  const configuration = getSampleSearchEngineConfiguration();
  configuration.search!.proxyBaseUrl = url;
  const engine = buildSearchEngine({
    configuration,
    navigatorContextProvider: () => ({
      clientId: crypto.randomUUID(),
      location: 'some location',
      referrer: 'some referrer',
      userAgent: 'some user agent',
    }),
  });
  return buildSearchBox(engine);
};

describe('SearchBox', () => {
  let provider: PactV3;
  beforeEach(() => {
    vi.spyOn(Math, 'random').mockReturnValue(0.5);
    provider = providerFactory();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('when there are suggestions available', () => {
    beforeEach(() => {
      provider = addSuggestionInteraction(provider);
    });

    it('should return suggestions', async () => {
      await provider.executeTest(async (mockServer) => {
        const searchBox = getSearchBoxController(mockServer.url);
        searchBox.showSuggestions();
        await waitForNSubscribeCallback(searchBox, 2);
        expect(searchBox.state).toMatchSnapshot();
      });
    });
  });
});

async function waitForNSubscribeCallback(
  searchBox: SearchBox,
  subscribeThreshold: number
) {
  await new Promise<void>((resolve) => {
    let subscribeCount = 0;
    searchBox.subscribe(() => {
      if (++subscribeCount < subscribeThreshold) {
        return;
      }
      resolve();
    });
  });
}
