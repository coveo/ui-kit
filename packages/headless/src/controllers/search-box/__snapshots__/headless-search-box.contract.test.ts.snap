// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBox > when there are suggestions available > should return suggestions 1`] = `
{
  "isLoading": false,
  "isLoadingSuggestions": false,
  "searchBoxId": "search_boxi",
  "suggestions": [
    {
      "highlightedValue": "suede",
      "rawValue": "suede",
    },
  ],
  "value": "",
}
`;
